### PR TITLE
[GEOS-11351] Exact term search in the pages' filters

### PR DESCRIPTION
**Pull request by axl8713**
[GEOS-11351](https://osgeo-org.atlassian.net/browse/GEOS-11351) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

These changes add the ability to refine element searches within paginated results through the search/filter input field, providing users with the option to request an exact-term filter by surrounding the term with quotes or double-quotes.
  

# Checklist

- [X] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [X] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [X] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [X] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [X] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/user/source/gettingstarted/web-admin-quickstart/index.rst [hunk 1]
```diff
@@ -73,6 +73,8 @@ The :ref:`layerpreview` page allows you to quickly view the output of published
    
       Search results for the query "top" on the Workspace page
 
+   .. hint:: Perform an exact term search by enclosing the search term in quotes or double-quotes, e.g. "topp".
```
Comment (aaime) [inline doc/en/user/source/gettingstarted/web-admin-quickstart/index.rst:76]:
```suggestion
   .. hint:: Perform an exact term search by enclosing the search term in quotes or double-quotes, e.g. normally ``ads`` would also match ``roads``, but ``"ads"`` wouldn't.
```

File: src/main/src/main/java/org/geoserver/catalog/Predicates.java [hunk 2]
```diff
@@ -198,6 +198,17 @@ public static Filter fullTextSearch(final String subsequence) {
...
 
+    /**
+     * Returns a predicate for exact match search (case-insensitive) against the <code>AnyText
+     * </code> property with the provided term.
+     *
+     * @param term the exact term to search for, ignoring case
+     * @return the filter for exact match search
+     */
+    public static Filter exactTermSearch(final String term) {
+        return factory.like(ANY_TEXT, fixSpecials(term));
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/Predicates.java:209]: Why use "like", it seems you're looking for simple equality... if so, ``Predicates.equal`` should also work.

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java [hunk 3]
```diff
@@ -3514,6 +3514,165 @@ public void testFullTextSearchAddedKeyword() {
...
+
+        LayerInfo lproxy = catalog.getLayer(l.getId());
+        FeatureTypeInfo ftproxy = (FeatureTypeInfo) lproxy.getResource();
+
+        ftproxy.getKeywords().add(new Keyword("newKeyword"));
+        catalog.save(ftproxy);
+
+        Filter filter = Predicates.exactTermSearch("newKeyword");
+        assertEquals(newHashSet(ftproxy), asSet(catalog.list(FeatureTypeInfo.class, filter)));
+        assertEquals(newHashSet(lproxy), asSet(catalog.list(LayerInfo.class, filter)));
```
Comment (aaime) [inline src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java:3643]: I would add a test before the keyword being added, which should result in an empty result

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java [hunk 4]
```diff
@@ -3514,6 +3514,165 @@ public void testFullTextSearchAddedKeyword() {
...
+        addLayer();
+        catalog.add(cv);
+
+        LayerInfo l2 = newLayer(cv, s);
+        catalog.add(l2);
+
+        Filter filter = Predicates.exactTermSearch("water");
+        assertEquals(newHashSet(ft), asSet(catalog.list(ResourceInfo.class, filter)));
+        assertEquals(newHashSet(ft), asSet(catalog.list(FeatureTypeInfo.class, filter)));
+        assertTrue(asSet(catalog.list(CoverageInfo.class, filter)).isEmpty());
```
Comment (aaime) [inline src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java:3662]: In general, for new code, we try to avoid using assertTrue/assertFalse. Hamcrest provides a way to make the test fail with a more informative message, when that happens (assertTrue/assertFalse only report "I was expecting true but got false", Hamcrest reports "I was expecting an empty collection but got a, b, c")

```suggestion
        assertThat(asSet(catalog.list(CoverageInfo.class, filter)), Matchers.empty());
```

File: src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerDataProvider.java [hunk 5]
```diff
@@ -43,6 +41,8 @@
  * @param <T>
  */
 public abstract class GeoServerDataProvider<T> extends SortableDataProvider<T, Object> {
+    public static final Pattern EXACT_TERM_KEYWORD_PATTERN =
+            Pattern.compile("^([\"'])(?<keyword>.+)\\1$");
```
Comment (aaime) [inline src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerDataProvider.java:45]: This one needs some explanation, something like:

```
// matches exact terms enclosed by either double or single quotes
// The first matching group matches either " or ', the "keyword" matching group matches any sequence of at least one character, the last group refers to the first one (\\1$) ensuring the start and end quote are balanced
```
A few more notes:
* Does the field need to be public? If not, default to private or protected.
* The documentation does not mention single quote being a possibility, please amend either docs or implementation, or people will wonder why searching for ``un'azione dell'altro mondo`` weirdly only matches ``azione dell`` but not ``azione``. Given the single quote has legit usages, I'd go for removing it from the regex, which will also simplify it
* I don't see a way to escape the double quote character, should it happen to be a valid value in the string. This could also be addressed by documentation, stating explicitly that there is no escaping support.

File: src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerTablePanel.html [hunk 6]
```diff
@@ -12,6 +12,7 @@
 		<div class="float-end">
 			<input id="filter" wicket:id="filter" class="me-1 search" type="text" size="35" title="Search" placeholder="Search" wicket:message="placeholder:Search"/>
 			<input wicket:id="submit" class="d-none" type="submit" value="Search"/>
+			<span class="help-link" title="Enclose the search term in quotes for exact-term search."></span>
```
Comment (aaime) [inline src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerTablePanel.html:15]: This should be subject to internationalization (you can provide support for english only for starters, in ``GeoServerApplication.properties``). Also, make sure it alignes with actual implementation (single or double quotes, see above).

File: src/web/core/src/test/java/org/geoserver/web/wicket/GeoServerDataProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                }
+            };
+
+    @Test
+    public void fullTextFilter() {
+        geoServerDataProvider.setKeywords(new String[] {"keyword"});
+
+        Filter filter = geoServerDataProvider.getFilter();
+
+        assertEquals("[ AnyText is like *keyword* ]", filter.toString());
```
Comment (aaime) [inline src/web/core/src/test/java/org/geoserver/web/wicket/GeoServerDataProviderTest.java:37]: Testing against toString is unreliable, developers are free to change its result without changing the semantics of the method (any string that "textually represents the object" is valid). Instead, use ``ECQL.toCQL(filter)``, it provides a text output with a format and stable grammar.

I'm not commenting the other cases in this class, but please fix them all.

**Review by aaime — CHANGES_REQUESTED** [review]
Overall looks good, see comments below.

File: src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerDataProvider.java (hunk 5 above)
Comment (axl8713) [reply src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerDataProvider.java:45]: Added a Javadoc for the field and made it private. Thanks for the feedback.

Regarding the search query behavior, if a query such as 'un'azione dell'altro mondo` is entered, the query will be tokenized by splitting on spaces, resulting in the search terms: `["un'azione", "dell'altro", "mondo"]`. This maintains the legitimate use of the single quote within the words.

Additionally, the double quote character could still be part of the string. For example, searching for `un'azione dell'altro "mondo` would produce the keywords: `["un'azione", "dell'altro", ""mondo"]`.

File: doc/en/user/source/gettingstarted/web-admin-quickstart/index.rst (hunk 1 above)
Comment (axl8713) [reply doc/en/user/source/gettingstarted/web-admin-quickstart/index.rst:76]: Added the example. Thanks.

File: src/web/core/src/test/java/org/geoserver/web/wicket/GeoServerDataProviderTest.java (hunk 7 above)
Comment (axl8713) [reply src/web/core/src/test/java/org/geoserver/web/wicket/GeoServerDataProviderTest.java]: Fixed, thanks.
Removed also other `toString()` in assertions.

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java (hunk 4 above)
Comment (axl8713) [reply src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java:3662]: Changed. Thanks for the feedback.

File: src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java (hunk 3 above)
Comment (axl8713) [reply src/main/src/test/java/org/geoserver/catalog/impl/CatalogImplTest.java:3643]: Added a test for a search miss on a different keyword. Thanks.

File: src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerTablePanel.html (hunk 6 above)
Comment (axl8713) [reply src/web/core/src/main/java/org/geoserver/web/wicket/GeoServerTablePanel.html]: Added. Thanks.

File: src/main/src/main/java/org/geoserver/catalog/Predicates.java (hunk 2 above)
Comment (aaime) [reply src/main/src/main/java/org/geoserver/catalog/Predicates.java:209]: Equal can also run in case-insensitive mode, but Predicates.equal is forcing case sensitiveness inside. Simple equality is faster, like is implemented as regular expressions, but it should not be a big deal. You can either expose a Predicates.equal variant that allows to tun the matchCase attribute, or keep it this as it is.

**Review by aaime — APPROVED** [review]
Guess the exact match performance can be optimized later, if needed. I guess it won't be the number one use case for the search bar anyways
